exonum-java-binding-cryptocurrency-demo/src/main/java/com/exonum/binding/cryptocurrency/CryptocurrencyServiceImpl.java

```diff
@@ -82,15 +89,39 @@ public void createPublicApiHandlers(Node node, Router router) {
 
     return node.withSnapshot(view -> {
       CryptocurrencySchema schema = new CryptocurrencySchema(view);
-      ListIndex<HistoryEntity> history = schema.walletHistory(ownerKey);
+      ListIndex<HashCode> walletHistory = schema.transactionsHistory(ownerKey);
+      Blockchain blockchain = Blockchain.newInstance(view);
+      MapIndex<HashCode, TransactionMessage> txMessages = blockchain.getTxMessages();
+
+      return stream(walletHistory)
+          .map(txMessages::get)
+          .map(this::createTransferHistoryEntry)
+          .collect(toList());
+    });
+  }
 
-      List<HistoryEntity> result = new ArrayList<>();
-      for (HistoryEntity tx : history) {
-        result.add(tx);
-      }
+  private static Stream<HashCode> stream(ListIndex<HashCode> list) {
+    // TODO: Replace with a proper thing in ECR-597!
+    return StreamSupport.stream(list.spliterator(), false);
+  }
 
-      return result;
-    });
+  private HistoryEntity createTransferHistoryEntry(TransactionMessage txMessage) {
+    checkState(txMessage.getServiceId() == getId(),
+        "service id mismatch: message contains %s, expected %s", txMessage.getServiceId(), getId());
+    try {
+      TxMessageProtos.TransferTx txBody = TxMessageProtos.TransferTx
```

Comment:
`StandardSerializers#protobuf`?

Reply:
I'd rather not — it requires a separate constant static field :-)